akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldResourceAsyncSinkSpec.scala

```diff
@@ -0,0 +1,246 @@
+        }
+        isClosed = true
+      }
+      if (autoCompleteClose) {
+        Future.successful(syncClose())
+      } else {
+        completeClose = Promise[Unit]
+        completeClose.future.map { _ ⇒
+          syncClose()
+        }
+      }
+    }
+
+  }
+
+  val settings = ActorMaterializerSettings(system).withDispatcher("akka.actor.default-dispatcher")
+  implicit val materializer = ActorMaterializer(settings)
+
+  val manyLines = {
+    ("a" * 100 + "\n") * 10 +
+      ("b" * 100 + "\n") * 10 +
+      ("c" * 100 + "\n") * 10 +
+      ("d" * 100 + "\n") * 10 +
+      ("e" * 100 + "\n") * 10 +
+      ("f" * 100 + "\n") * 10
+  }
+  val manyLinesArray = manyLines.split("\n")
+
+  "FoldResourceSinkAsync" must {
+    "call open, write, close in order" in assertAllStagesStopped {
+      case class DummyWriter()
+
+      val openPromise = Promise[DummyWriter]
+      val writePromise = Promise[Unit]
+      val closePromise = Promise[Unit]
+
+      val openCalled = Promise[Unit]
+      val writeCalled = Promise[Unit]
+      val closeCalled = Promise[Unit]
+
+      val sink = Sink.foldResourceAsync[String, DummyWriter](
+        () ⇒ {
+          openCalled.success(())
+          openPromise.future
+        }, (writer, s) ⇒ {
+          writeCalled.success(())
+          writePromise.future
+        }, writer ⇒ {
+          closeCalled.success(())
+          closePromise.future
+        }
+      )
+
+      val (probe, fut) = TestSource.probe[String].toMat(sink)(Keep.both).run()
+
+      Await.result(openCalled.future, remainingOrDefault)
+      openPromise.success(DummyWriter())
+
+      writeCalled.isCompleted should be(false)
+      probe.sendNext("A")
+      Await.result(writeCalled.future, remainingOrDefault)
+      writePromise.success(())
+
+      closeCalled.isCompleted should be(false)
+      fut.isCompleted should be(false)
+      probe.sendComplete()
+      Await.result(closeCalled.future, remainingOrDefault)
+      closePromise.success(())
+
+      Await.result(fut, remainingOrDefault)
+    }
+
+    "fail a materialized value when write throws an exception" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](
+        () ⇒ r.open(),
+        (w, s) ⇒ Future {
+          if (s.contains("b")) throw TE("")
+          else w.write(s)
+        },
+        _.close()
+      )
+
+      val fut = Source.fromIterator(() ⇒ manyLines.grouped(64)).runWith(sink)
+      Await.result(fut.failed, remainingOrDefault) should be(TE(""))
+      (r.isOpened, r.isClosed) should be((true, true))
+    }
+
+    "write contents to a resource" in assertAllStagesStopped {
+      val r = new TestResource[String]()
+      val sink = Sink.foldResourceAsync[String, TestResource[String]](() ⇒ r.open(), _ write _, _.close())
+      val fut = Source.fromIterator(() ⇒ manyLines.grouped(64)).runWith(sink)
+
+      Await.result(fut, remainingOrDefault) should be(Done)
```

Comment:
isn't there an implicit-decoration ".futureValue" available here to use instead of `Await.result`?